tests/attacks/evasion/test_rescaling_auto_conjugate_gradient.py

```diff
@@ -0,0 +1,166 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import logging
+import pytest
+
+import numpy as np
+
+from art.attacks.evasion import RescalingAutoConjugateGradient
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.classification.classifier import ClassifierMixin
+
+from tests.attacks.utils import backend_test_classifier_type_check_fail
+from tests.utils import ARTTestException
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture()
+def fix_get_mnist_subset(get_mnist_dataset):
+    (x_train_mnist, y_train_mnist), (x_test_mnist, y_test_mnist) = get_mnist_dataset
+    n_train = 100
+    n_test = 10
+    yield x_train_mnist[:n_train], y_train_mnist[:n_train], x_test_mnist[:n_test], y_test_mnist[:n_test]
+
+
+@pytest.mark.parametrize("loss_type", ["cross_entropy", "difference_logits_ratio"])
+@pytest.mark.parametrize("norm", ["inf", np.inf, 1, 2])
+@pytest.mark.skip_framework("keras", "non_dl_frameworks", "mxnet", "kerastf", "tensorflow1", "tensorflow2v1")
+def test_generate(art_warning, fix_get_mnist_subset, image_dl_estimator_for_attack, framework, loss_type, norm):
+    print("test_generate")
+    try:
+        classifier = image_dl_estimator_for_attack(RescalingAutoConjugateGradient, from_logits=True)
+
+        print("framework", framework)
+
+        if framework in ["tensorflow1", "tensorflow2v1"] and loss_type == "difference_logits_ratio":
+            with pytest.raises(ValueError):
+                _ = RescalingAutoConjugateGradient(
+                    estimator=classifier,
+                    norm=norm,
+                    eps=0.3,
+                    eps_step=0.1,
+                    max_iter=5,
+                    targeted=False,
+                    nb_random_init=1,
+                    batch_size=32,
+                    loss_type=loss_type,
+                    verbose=False,
+                )
+        else:
+
+            attack = RescalingAutoConjugateGradient(
+                estimator=classifier,
+                norm=norm,
+                eps=0.3,
+                eps_step=0.1,
+                max_iter=5,
+                targeted=False,
+                nb_random_init=1,
+                batch_size=32,
+                loss_type=loss_type,
+                verbose=False,
+            )
+
+            (x_train_mnist, y_train_mnist, x_test_mnist, y_test_mnist) = fix_get_mnist_subset
+
+            x_train_mnist_adv = attack.generate(x=x_train_mnist, y=y_train_mnist)
+
+            assert np.max(np.abs(x_train_mnist_adv - x_train_mnist)) > 0.0
```

Comment:
Does the attack always create the same adversarial example? Could we also test for the expected pixel values in the adversarial image?

Reply:
A CPU-only calculation with the same initial point should produce the same adversarial example each time.
However, when using a GPU, the adversarial example may be slightly different each time because the gradient calculation on the GPU is not deterministic.